mmengine/hooks/early_stopping_hook.py

```diff
@@ -0,0 +1,98 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import warnings
+from typing import List, Optional, Union
+
+from mmengine.registry import HOOKS
+from .hook import Hook
+
+DATA_BATCH = Optional[Union[dict, tuple, list]]
+
+
+@HOOKS.register_module()
+class EarlyStoppingHook(Hook):
+    """Early stop the training when the metric reached a plateau.
+
+    Args:
+        metric (str): The metric key to decide early stopping.
+        rule (str, optional): Comparison rule. Options are 'greater',
+            'less'. Defaults to None.
+        delta(float, optional): Minimum difference to continue the training.
+            Defaults to 0.01.
+        pool_size (int, optional): The number of experiments to consider.
+            Defaults to 5.
+        patience (int, optional): Maximum number of tolerance.
+            Defaults to 0.
+    """
+    priority = 'LOWEST'
+
+    rule_map = {'greater': lambda x, y: x > y, 'less': lambda x, y: x < y}
+    _default_greater_keys = [
+        'acc', 'top', 'AR@', 'auc', 'precision', 'mAP', 'mDice', 'mIoU',
+        'mAcc', 'aAcc'
+    ]
+    _default_less_keys = ['loss']
+
+    def __init__(
+        self,
+        metric: str,
+        rule: str = None,
+        delta: float = 0.1,
+        pool_size: int = 5,
+        patience: int = 0,
+    ):
+
+        self.metric = metric
+        if metric in self._default_greater_keys:
+            rule = 'greater'
+        elif metric in self._default_less_keys:
+            rule = 'less'
+        assert rule in ['greater', 'less'], \
+            '`rule` should be either \'greater\' or \'less\'.'
+        self.rule = rule
+        self.delta = delta
+        self.pool_size = pool_size
+        self.patience = patience
+        self.count = 0
+
+        self.pool_values: List[float] = []
+
+    def after_val_epoch(self, runner, metrics):
+        """Decide whether to stop the training process.
+
+        Args:
+            runner (Runner): The runner of the training process.
+            metrics (dict): Evaluation results of all metrics
+        """
+
+        if self.metric not in metrics:
+            warnings.warn(
+                f'Skip early stopping process since the evaluation '
+                f'results ({metrics.keys()}) do not include `metric` '
+                f'({self.metric}).')
+            return
+
+        latest_value = metrics[self.metric]
+        compare = self.rule_map[self.rule]
+
+        self.pool_values.append(latest_value)
+
+        if self.rule == 'greater':
+            # maintain largest values
+            self.pool_values = sorted(self.pool_values)[-self.pool_size:]
+        else:
+            # maintain smalleast values
+            self.pool_values = sorted(self.pool_values)[:self.pool_size]
```

Comment:
How about using `PriorityQueue`?

Reply:
https://stackoverflow.com/a/26441285
If `rule=='greater`, getting values from the `PriorityQueue` should work differently. `Q.get()` or `Q.get()[1]`.
Is there any good idea to handle this case?

Reply:
Initially, I thought we can put + or - metrics in `PriorityQueue` to represent `greater` or `less`. Emmmm. But I think since the metrics list will not be too long, it could be OK to directly use `list` rather than `PriorityQueue`